scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkBackend.scala

```diff
@@ -0,0 +1,226 @@
+    Memoize.functionK(new Memoize.RecursiveK[TypedPipe, R] {
+      import TypedPipe._
+
+      def toFunction[A] = {
+        case (cp @ CounterPipe(_), rec) =>
+          // TODO: counters not yet supported
+          def go[A](p: CounterPipe[A]): R[A] = {
+            rec(p.pipe).map(_._1)
+          }
+          go(cp)
+        case (cp @ CrossPipe(_, _), rec) =>
+          def go[A, B](cp: CrossPipe[A, B]): R[(A, B)] =
+            rec(cp.viaHashJoin)
+          go(cp)
+        case (CrossValue(left, EmptyValue), rec) => rec(EmptyTypedPipe)
+        case (CrossValue(left, LiteralValue(v)), rec) =>
+          val op = rec(left) // linter:disable:UndesirableTypeInference
+          op.map((_, v))
+        case (CrossValue(left, ComputedValue(right)), rec) =>
+          rec(CrossPipe(left, right))
+        case (p: DebugPipe[a], rec) =>
+          // There is really little that can be done here but println
+          rec[a](p.input.map(DebugFn()))
+        case (EmptyTypedPipe, rec) =>
+          R.empty(ctx)
+        case (fk @ FilterKeys(_, _), rec) =>
+          def go[K, V](node: FilterKeys[K, V]): R[(K, V)] = {
+            val FilterKeys(pipe, fn) = node
+            rec(pipe).filter(FilterKeysToFilter(fn))
+          }
+          go(fk)
+        case (f @ Filter(_, _), rec) =>
+          def go[T](f: Filter[T]): R[T] = {
+            val Filter(p, fn) = f
+            rec[T](p).filter(fn)
+          }
+          go(f)
+        case (f @ FlatMapValues(_, _), rec) =>
+          def go[K, V, U](node: FlatMapValues[K, V, U]) = {
+            val fn = node.fn
+            rec(node.input).flatMapValues(fn)
+          }
+
+          go(f)
+        case (FlatMapped(prev, fn), rec) =>
+          val op = rec(prev) // linter:disable:UndesirableTypeInference
+          op.concatMap(fn)
+        case (ForceToDisk(pipe), rec) =>
+          rec(pipe).persist(StorageLevel.DISK_ONLY)
+        case (Fork(pipe), rec) =>
+          rec(pipe).persist(StorageLevel.MEMORY_ONLY)
+        case (IterablePipe(iterable), _) =>
+          R.fromIterable(ctx, iterable)
+        case (f @ MapValues(_, _), rec) =>
+          def go[K, V, U](node: MapValues[K, V, U]): R[(K, U)] =
+            rec(node.input).mapValues(node.fn)
+          go(f)
+        case (Mapped(input, fn), rec) =>
+          val op = rec(input) // linter:disable:UndesirableTypeInference
+          op.map(fn)
+        case (m @ MergedTypedPipe(_, _), rec) =>
+          def go[A](m: MergedTypedPipe[A]): R[A] = {
+            R.merge(rec(m.left), rec(m.right))
+          }
+          go(m)
+        case (SourcePipe(src), rec) =>
+          // TODO
+          ???
+        case (slk @ SumByLocalKeys(_, _), rec) =>
+          def sum[K, V](sblk: SumByLocalKeys[K, V]): R[(K, V)] = {
+            // we can use Algebird's SummingCache https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/SummingCache.scala#L36
```

Comment:
since these partitions are ~usually ondisk or can fit in memory we might just want MapAlgebra.sumByKey possibly (or sort and fold i guess..).